src/hdf5_dl.c

```diff
@@ -97,44 +97,167 @@ hid_t H5E_ERR_CLS_g = -1;
 
 static bool is_init = false;
 
+/*
+ * Try to find a symbol within a library
+ * 
+ * handle: Handle to the library
+ * symbol: Symbol to look for
+ * Returns: a pointer to the symbol or NULL 
+ * if the symbol can't be found 
+ */
+void *find_sym(void *handle, const char *symbol) {
+
+  void *ret = NULL, *err = NULL;
+  dlerror(); /* clear error code */
+  ret = dlsym(handle, symbol);
+
+  if(ret != NULL && (err = dlerror()) == NULL)
+    return ret;
+  else
+    return NULL;
+}
 
 /* Initialize the dynamic loading of symbols and register the plugin
  *
  * libname: Name of the DLL from which to load libHDF5 symbols
  * Returns: a value < 0 if an error occured
  */
-int init_filter(const char* libname)
+int init_filter(const char *libname)
 {
     int retval = -1;
-  	void * handle;
+  	void *handle = NULL, *symbol = NULL;
 
     handle = dlopen(libname, RTLD_LAZY | RTLD_LOCAL);
 
     if (handle != NULL) {
         /*H5*/
-        DL_H5Functions.H5open = (DL_func_H5open)dlsym(handle, "H5open");
+        if(DL_H5Functions.H5open == NULL) {
```

Comment:
These repeated blocks were a little hard to follow. It seems there are two things being done at the same time which prevent things being more neatly structured:
- Attempt to resolve a symbol if it is not already resolved
- Check if all symbols are now resolved

I can see it being a lot cleaner to completely separate these, i.e. you have one pass doing the first, and then after you've done that an entirely separate pass (probably placed in a separate function) which checks that all the symbols have been resolved.